client/components/swipeable/index.js

```diff
@@ -0,0 +1,212 @@
+			};
+		}, [ enabled ] );
+	}
+
+	const updateEnabled = hasDynamicHeight && numPages > 1;
+
+	useUpdateLayout( updateEnabled, currentPage, ( hasTransitionDuration = true ) => {
+		const offset = getOffset( currentPage );
+		const transform = {
+			transform: `translate3d(${ offset }px, 0px, 0px)`,
+			transitionDuration: hasTransitionDuration ? `300ms` : `0ms`,
+		};
+
+		let height = { height: null };
+		const targetHeight = pagesRef.current?.querySelector( '.is-current' )?.offsetHeight;
+
+		if ( targetHeight && pagesStyle?.height !== targetHeight ) {
+			height = { height: null };
+		}
+
+		setPagesStyle( { ...pagesStyle, ...height, ...transform } );
+	} );
+
+	useEffect( () => {
+		if ( currentPage >= numPages ) {
+			onPageSelect( numPages - 1 );
+		}
+	}, [ numPages, currentPage, onPageSelect ] );
+
+	const getDragPosition = ( event ) => {
+		if ( event.hasOwnProperty( 'clientX' ) ) {
+			return { x: event.clientX, y: event.clientY };
+		}
+
+		if ( event.targetTouches[ 0 ] ) {
+			return { x: event.targetTouches[ 0 ].clientX, y: event.targetTouches[ 0 ].clientY };
+		}
+
+		const touch = event.changedTouches[ 0 ];
+		return { x: touch.clientX, y: touch.clientY };
+	};
+
+	const handleDragStart = ( event ) => {
+		const position = getDragPosition( event );
+		setDragStartData( position );
+		setPagesStyle( { ...pagesStyle, transitionDuration: `0ms` } ); // Set transition Duration to 0 for smooth dragging.
+		setIsDragging( true );
+	};
+
+	const handleDragEnd = ( event ) => {
+		if ( ! isDragging ) {
+			return; // End early if we are not dragging any more.
+		}
+
+		const THRESHOLD_OFFSET = 100; // Number of pixels to travel before we trigger the slider to move to the desired slide.
+		const dragPosition = getDragPosition( event );
+		const delta = dragPosition.x - dragStartData.x;
+
+		const hasMetThreshold = Math.abs( delta ) > THRESHOLD_OFFSET;
+
+		let newIndex = currentPage;
+		if ( delta < 0 && hasMetThreshold && numPages !== currentPage + 1 ) {
+			newIndex = currentPage + 1;
+		}
+
+		if ( delta > 0 && hasMetThreshold && currentPage !== 0 ) {
+			newIndex = currentPage - 1;
+		}
+		const offset = getOffset( newIndex );
+		setPagesStyle( {
+			transform: `translate3d(${ offset }px, 0px, 0px)`,
+			transitionDuration: `300ms`,
+		} );
+		onPageSelect( newIndex );
+		setDragStartData( {} );
+		setIsDragging( false );
+	};
+
+	const handleDrag = ( event ) => {
+		if ( ! isDragging ) {
+			return;
+		}
+		const dragPosition = getDragPosition( event );
+		const delta = dragPosition.x - dragStartData.x;
+
+		const offset = getOffset( currentPage ) + delta;
+
+		// Allow for swipe left or right
+		if ( ( numPages !== currentPage + 1 && delta < 0 ) || ( currentPage !== 0 && delta > 0 ) ) {
+			setPagesStyle( {
+				...pagesStyle,
+				transform: `translate3d(${ offset }px, 0px, 0px)`,
+				transitionDuration: `0ms`,
+			} );
+		}
+
+		const swipeableArea = pagesRef.current?.getBoundingClientRect();
```

Comment:
Similarly to what state above, calling `getBoundingClientRect` on every drag event can hurt rendering performance. Ideally the size of the swipeable area should be a memoized value that is recalculated only when necessary

Reply:
Good call!